compiler/damlc/lib/DA/Cli/Damlc/Test.hs

```diff
@@ -151,6 +186,80 @@ testRun h inFiles lfVersion (RunAllTests runAllTests) coverage color mbJUnitOutp
                     pure (file, map (second render) scenarioResults)
         writeFile junitOutput $ XML.showTopElement $ toJUnit res
 
+data NamedPath = NamedPath { np_name :: String, np_path :: FilePath }
+    deriving (Show, Eq, Ord)
+
+tryWithPath :: (FilePath -> IO a) -> NamedPath -> IO (Maybe a)
+tryWithPath action NamedPath {..} =
+    handleJust errorFilter handler (Just <$> action np_path)
+    where
+        errorFilter :: IOError -> Maybe IOError
+        errorFilter err = do
+            guard $ isPermissionError err || isAlreadyExistsError err || isDoesNotExistError err
+            pure err
+
+        handler :: IOError -> IO (Maybe a)
+        handler err
+          | isPermissionError err = do
+              hPutStrLn stderr $ np_name ++ " cannot be created because of unsufficient permissions."
+              pure Nothing
+          | isAlreadyExistsError err = do
+              hPutStrLn stderr $ np_name ++ " cannot be created because it already exists."
+              pure Nothing
+          | isDoesNotExistError err = do
+              hPutStrLn stderr $ np_name ++ " cannot be created because its parent directory does not exist."
+              pure Nothing
+          | otherwise = throwIO err
```

Comment:
small suggestion: to avoid using the `is*` functions twice, we can change `errorFilter` to `:: IOError -> Maybe String`, returning your error strings, and then `handler` to `:: String -> IO (Maybe a)`, just `hPutStrLn stderr`ing its argument and returning `Nothing`

Reply:
Yes, I like that! That bit of code was bothering me because it looked very refactorable but I couldn't quite put my finger on what it was :P